semgrep-core/src/fixing/Autofix_printer.ml

```diff
@@ -0,0 +1,128 @@
+(* Nat Mote
+ *
+ * Copyright (C) 2019-2022 r2c
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+open Common
+open AST_generic
+module MV = Metavariable
+
+(******************************************************************************)
+(* Handles AST printing for the purposes of autofix.
+ *
+ * The main printing logic happens elsewhere. This module's main purpose is to
+ * extend the existing AST printers so that they can avoid printing AST nodes
+ * which are lifted unchanged from either the target file (via metavariable
+ * bindings) or the fix pattern in the rule. This serves two purposes:
+ * - It lets us synthesize autofixes without having to implement printing for
+ *   all of the AST nodes in the fixed AST.
+ * - It lets us make minimal changes to target files by carrying over
+ *   formatting, comments, etc. from the original code.
+ *)
+(******************************************************************************)
+
+module PythonPrinter = Hybrid_print.Make (struct
+  class printer = Ugly_print_AST.python_printer
+end)
+
+let get_printer lang external_printer : Ugly_print_AST.printer_t option =
+  match lang with
+  | Lang.Python -> Some (new PythonPrinter.printer external_printer)
+  | _ -> None
+
+let original_source_of_ast source any =
+  let* start, end_ = Visitor_AST.range_of_any_opt any in
```

Comment:
Note: this is fine now but if we ever start doing autofixes on stmt lists we could run into perf trouble. Just making this note because we've had multiple perf problems with `range_of_any_opt` and it's easy to run into. Getting a range of an expression or a statement is safer because those two are now at least cached.

Reply:
I'm hoping that, when compared to the overall amount of code scanned, the number of actual findings with fixes is low enough that we can get away with some inefficiency here. I read the comments on perf and I guessed that this was only an issue because it was getting called in a very hot code path. I don't think this is going to get called a ton.

I agree we should keep an eye on it, though.